lec05-dplyr.Rmd

```diff
@@ -1,4 +1,511 @@
+
+```{r}
+surveys_gw_wide %>%
+  gather(genus, mean_weight, Baiomys:Spermophilus) %>%
+  head()
+```
+
+># ### Challenge
+>
+> 1. Make a wide data frame with `year` as columns, `plot_id` as rows, and the
+>   values are the number of genera per plot. You will need to summarize before
+>   reshaping, and use the function `n_distinct` to get the number of unique
+>   types of a genus. It's a powerful function! See `?n_distinct` for more.
+>
+> 2. Now take that data frame, and make it long again, so each row is a unique
+>    `plot_id` - `year` combination.
+>
+> 3. The `surveys` data set is not truly wide or long because there are
+>    two columns of measurement - `hindfoot_length` and `weight`.  This makes it
+>    difficult to do things like look at the relationship between mean values of
+>    each measurement per year in different plot types. Let's walk through a
+>    common solution for this type of problem. First, use `gather` to create a
+>    truly long dataset where we have a key column called `measurement` and a
+>    `value` column that takes on the value of either `hindfoot_length` or
+>    `weight`. Hint: You'll need to specify which columns are being gathered.
+>
+> 4. With this new truly long data set, calculate the average of each
+>    `measurement` in each `year` for each different `plot_type`. Then
+>    `spread` them into a wide data set with a column for `hindfoot_length` and
+>    `weight`. Hint: Remember, you only need to specify the key and value
+>    columns for `spread`.
+
+<!---
+  ```{r, echo=FALSE}
+## Answer 1
+rich_time <- surveys %>%
+  group_by(plot_id, year) %>%
+  summarize(n_genera = n_distinct(genus)) %>%
+  spread(year, n_genera)
+
+head(rich_time)
+
+## Answer 2
+rich_time %>%
+  gather(year, n_genera, -plot_id)
+
+## Answer 3
+surveys_long <- surveys %>%
+  gather(measurement, value, hindfoot_length, weight)
+
+## Answer 4
+surveys_long %>%
+  group_by(year, measurement, plot_type) %>%
+  summarize(mean_value = mean(value, na.rm=TRUE)) %>%
+  spread(measurement, mean_value)
+```
+--->
+
+## Exporting data
+
+Now that you have learned how to use **`dplyr`** to extract information from
+or summarize your raw data, you may want to export these new datasets to share
+them with your collaborators or for archival.
+
+Similar to the `read_csv()` function used for reading CSV files into R, there is
+a `write_csv()` function that generates CSV files from data frames.
+
+Before using `write_csv()`, we are going to create a new folder, `data`,
+in our working directory that will store this generated dataset. We don't want
+to store manipulated datasets in the same directory as our raw data. It's good
+practice to keep them separate. The `data` folder should only contain the raw,
+unaltered data, and should be left alone to make sure we don't delete or modify
+it. In contrast, our R code will generate the contents of the `data`
+directory, so even if the files it contains are deleted, we can always
+re-generate them.
```

Comment:
Yea, @QuLogic is right, this is inconsistent. I think what you mean is that the raw data should be in the `data-raw` folder (this is commonly named in R packages) and the usable/slightly modified data is in the `data` folder.

Reply:
Wasn't it good the way it was initially? Where the output was in a separate folder?

Ideally, I would probably have both `data-raw`, and maybe `data-processed` (data here is output, but might be used as input for something else), leaving the top level clearner.

Reply:
I imagine you will talk more in detail about these things during the project setup lecture?

Reply:
Yes I will. While there are many many ways of naming folder etc in research projects, I'd prefer to stick with the R package style of doing it since it is well established and known. Plus there are tons of useful tutorials/books out there on it (e.g. http://r-pkgs.had.co.nz/).

So you would have a file structure like:

```
./
|- data/
|- data-raw/
```

Reply:
data-raw would be the untouched data, and the data folder would be the 'tidied' up data

Reply:
@lwjohnst86 Can you merge and upload to the website when you get here? Would be good to show the team challenges for the second half.

Reply:
Actually would be good to have it in the beginning so I will merge now and we can adjust later.